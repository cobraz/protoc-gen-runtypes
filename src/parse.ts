import { AnyType, RootType } from 'generate-runtypes';
import { parse } from 'protocol-buffers-schema';
import type { Field, Message, Schema } from 'protocol-buffers-schema/types';

export function parseTypeFromSchema(
  field: Field,
  message: Message,
  protobufSchema: Schema,
): AnyType {
  const childEnum =
    message.enums.find((_enum) => _enum.name === field.type) ||
    protobufSchema.enums.find((_enum) => _enum.name === field.type);
  const childMessage =
    message.messages.find((_message) => _message.name === field.type) ||
    protobufSchema.messages.find((message) => message.name === field.type);

  if (childEnum) {
    return {
      kind: 'union',
      types: Object.keys(childEnum.values).map((v) => ({
        kind: 'literal',
        value: v,
      })),
    };
  }

  if (childMessage) {
    return {
      kind: 'named',
      name: childMessage.name,
    };
  }

  switch (field.type) {
    case 'bool':
      return {
        kind: 'boolean',
      };

    case 'int32':
    case 'sint32':
    case 'int64':
    case 'sint64':
    case 'float':
    case 'double':
    case 'uint64':
    case 'uint32':
      return {
        kind: 'number',
      };

    case 'bytes':
      return {
        kind: 'unknown',
      };

    case 'string':
      return {
        kind: 'string',
      };

    case 'map':
      return {
        kind: 'unknown',
      };

    default:
      return {
        kind: 'unknown',
      };
  }
}

export function parseToGenerator(buffer: string | Buffer): RootType[] {
  const schema = parse(buffer);

  return schema.messages.map((message) => {
    return {
      name: message.name,
      type: {
        kind: 'record',
        fields: message.fields.map((field) => ({
          name: field.name,
          type: parseTypeFromSchema(field, message, schema),
          export: true,
          nullable: !field.required,
        })),
      },
    };
  });
}
