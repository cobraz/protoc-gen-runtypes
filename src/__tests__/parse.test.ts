import { readFile } from 'fs/promises';
import { join } from 'path';
import * as camelCase from 'camelcase';
import { generateRuntypes } from 'generate-runtypes';
import { parseToGenerator } from '../parse';

const FIXTURES_DIR = join(__dirname, '..', '__fixtures__');

describe('parse', () => {
  describe('parseToGenerator', () => {
    it('should print out generate-runtypes schema from proto2', async () => {
      const proto = await readFile(join(FIXTURES_DIR, 'example-proto2.proto'));
      const schema = parseToGenerator(proto);

      expect(JSON.stringify(schema, null, 2)).toMatchSnapshot();

      const generated = await generateRuntypes(schema);

      expect(generated).toMatchSnapshot();
    });
    it('should print out generate-runtypes schema from proto3', async () => {
      const proto = await readFile(join(FIXTURES_DIR, 'example-proto3.proto'));
      const schema = parseToGenerator(proto);

      expect(JSON.stringify(schema, null, 2)).toMatchSnapshot();

      const generated = await generateRuntypes(schema, {
        formatRuntypeName: (n) => camelCase(n, { pascalCase: true }),
      });

      expect(generated).toMatchSnapshot();
    });
  });
});
