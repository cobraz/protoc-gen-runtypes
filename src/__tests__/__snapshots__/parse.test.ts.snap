// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`parse parseToGenerator should print out generate-runtypes schema from proto2 1`] = `
"[
  {
    \\"name\\": \\"Invoice\\",
    \\"type\\": {
      \\"kind\\": \\"record\\",
      \\"fields\\": [
        {
          \\"name\\": \\"name\\",
          \\"type\\": {
            \\"kind\\": \\"string\\"
          },
          \\"export\\": true,
          \\"nullable\\": true
        },
        {
          \\"name\\": \\"description\\",
          \\"type\\": {
            \\"kind\\": \\"string\\"
          },
          \\"export\\": true,
          \\"nullable\\": true
        },
        {
          \\"name\\": \\"status\\",
          \\"type\\": {
            \\"kind\\": \\"union\\",
            \\"types\\": [
              {
                \\"kind\\": \\"literal\\",
                \\"value\\": \\"INVOICE_STATUS_UNSPECIFIED\\"
              },
              {
                \\"kind\\": \\"literal\\",
                \\"value\\": \\"INVOICE_STATUS_PENDING\\"
              },
              {
                \\"kind\\": \\"literal\\",
                \\"value\\": \\"INVOICE_STATUS_SENT\\"
              },
              {
                \\"kind\\": \\"literal\\",
                \\"value\\": \\"INVOICE_STATUS_AWAITING_PAYMENT\\"
              },
              {
                \\"kind\\": \\"literal\\",
                \\"value\\": \\"INVOICE_STATUS_PAID\\"
              },
              {
                \\"kind\\": \\"literal\\",
                \\"value\\": \\"INVOICE_STATUS_DEPT_COLLECTION\\"
              }
            ]
          },
          \\"export\\": true,
          \\"nullable\\": true
        }
      ]
    }
  },
  {
    \\"name\\": \\"Point\\",
    \\"type\\": {
      \\"kind\\": \\"record\\",
      \\"fields\\": [
        {
          \\"name\\": \\"x\\",
          \\"type\\": {
            \\"kind\\": \\"number\\"
          },
          \\"export\\": true,
          \\"nullable\\": true
        },
        {
          \\"name\\": \\"y\\",
          \\"type\\": {
            \\"kind\\": \\"number\\"
          },
          \\"export\\": true,
          \\"nullable\\": true
        },
        {
          \\"name\\": \\"label\\",
          \\"type\\": {
            \\"kind\\": \\"string\\"
          },
          \\"export\\": true,
          \\"nullable\\": true
        }
      ]
    }
  },
  {
    \\"name\\": \\"Line\\",
    \\"type\\": {
      \\"kind\\": \\"record\\",
      \\"fields\\": [
        {
          \\"name\\": \\"start\\",
          \\"type\\": {
            \\"kind\\": \\"named\\",
            \\"name\\": \\"Point\\"
          },
          \\"export\\": true,
          \\"nullable\\": true
        },
        {
          \\"name\\": \\"end\\",
          \\"type\\": {
            \\"kind\\": \\"named\\",
            \\"name\\": \\"Point\\"
          },
          \\"export\\": true,
          \\"nullable\\": true
        },
        {
          \\"name\\": \\"label\\",
          \\"type\\": {
            \\"kind\\": \\"string\\"
          },
          \\"export\\": true,
          \\"nullable\\": true
        }
      ]
    }
  }
]"
`;

exports[`parse parseToGenerator should print out generate-runtypes schema from proto2 2`] = `
"import * as rt from \\"runtypes\\";

const invoice = rt
  .Record({
    name: rt.String,
    description: rt.String,
    status: rt.Union(
      rt.Literal(\\"INVOICE_STATUS_UNSPECIFIED\\"),
      rt.Literal(\\"INVOICE_STATUS_PENDING\\"),
      rt.Literal(\\"INVOICE_STATUS_SENT\\"),
      rt.Literal(\\"INVOICE_STATUS_AWAITING_PAYMENT\\"),
      rt.Literal(\\"INVOICE_STATUS_PAID\\"),
      rt.Literal(\\"INVOICE_STATUS_DEPT_COLLECTION\\")
    ),
  })
  .asPartial();

type Invoice = rt.Static<typeof invoice>;

const point = rt
  .Record({ x: rt.Number, y: rt.Number, label: rt.String })
  .asPartial();

type Point = rt.Static<typeof point>;

const line = rt
  .Record({ start: point, end: point, label: rt.String })
  .asPartial();

type Line = rt.Static<typeof line>;
"
`;

exports[`parse parseToGenerator should print out generate-runtypes schema from proto3 1`] = `
"[
  {
    \\"name\\": \\"Invoice\\",
    \\"type\\": {
      \\"kind\\": \\"record\\",
      \\"fields\\": [
        {
          \\"name\\": \\"name\\",
          \\"type\\": {
            \\"kind\\": \\"string\\"
          },
          \\"export\\": true,
          \\"nullable\\": false
        },
        {
          \\"name\\": \\"description\\",
          \\"type\\": {
            \\"kind\\": \\"string\\"
          },
          \\"export\\": true,
          \\"nullable\\": false
        },
        {
          \\"name\\": \\"status\\",
          \\"type\\": {
            \\"kind\\": \\"union\\",
            \\"types\\": [
              {
                \\"kind\\": \\"literal\\",
                \\"value\\": \\"INVOICE_STATUS_UNSPECIFIED\\"
              },
              {
                \\"kind\\": \\"literal\\",
                \\"value\\": \\"INVOICE_STATUS_PENDING\\"
              },
              {
                \\"kind\\": \\"literal\\",
                \\"value\\": \\"INVOICE_STATUS_SENT\\"
              },
              {
                \\"kind\\": \\"literal\\",
                \\"value\\": \\"INVOICE_STATUS_AWAITING_PAYMENT\\"
              },
              {
                \\"kind\\": \\"literal\\",
                \\"value\\": \\"INVOICE_STATUS_PAID\\"
              },
              {
                \\"kind\\": \\"literal\\",
                \\"value\\": \\"INVOICE_STATUS_DEPT_COLLECTION\\"
              }
            ]
          },
          \\"export\\": true,
          \\"nullable\\": true
        }
      ]
    }
  },
  {
    \\"name\\": \\"Point\\",
    \\"type\\": {
      \\"kind\\": \\"record\\",
      \\"fields\\": [
        {
          \\"name\\": \\"x\\",
          \\"type\\": {
            \\"kind\\": \\"number\\"
          },
          \\"export\\": true,
          \\"nullable\\": false
        },
        {
          \\"name\\": \\"y\\",
          \\"type\\": {
            \\"kind\\": \\"number\\"
          },
          \\"export\\": true,
          \\"nullable\\": false
        },
        {
          \\"name\\": \\"label\\",
          \\"type\\": {
            \\"kind\\": \\"string\\"
          },
          \\"export\\": true,
          \\"nullable\\": true
        }
      ]
    }
  },
  {
    \\"name\\": \\"Line\\",
    \\"type\\": {
      \\"kind\\": \\"record\\",
      \\"fields\\": [
        {
          \\"name\\": \\"start\\",
          \\"type\\": {
            \\"kind\\": \\"named\\",
            \\"name\\": \\"Point\\"
          },
          \\"export\\": true,
          \\"nullable\\": false
        },
        {
          \\"name\\": \\"end\\",
          \\"type\\": {
            \\"kind\\": \\"named\\",
            \\"name\\": \\"Point\\"
          },
          \\"export\\": true,
          \\"nullable\\": false
        },
        {
          \\"name\\": \\"label\\",
          \\"type\\": {
            \\"kind\\": \\"string\\"
          },
          \\"export\\": true,
          \\"nullable\\": true
        }
      ]
    }
  }
]"
`;

exports[`parse parseToGenerator should print out generate-runtypes schema from proto3 2`] = `
"import * as rt from \\"runtypes\\";

const Invoice = rt.Intersect(
  rt.Record({ name: rt.String, description: rt.String }),
  rt
    .Record({
      status: rt.Union(
        rt.Literal(\\"INVOICE_STATUS_UNSPECIFIED\\"),
        rt.Literal(\\"INVOICE_STATUS_PENDING\\"),
        rt.Literal(\\"INVOICE_STATUS_SENT\\"),
        rt.Literal(\\"INVOICE_STATUS_AWAITING_PAYMENT\\"),
        rt.Literal(\\"INVOICE_STATUS_PAID\\"),
        rt.Literal(\\"INVOICE_STATUS_DEPT_COLLECTION\\")
      ),
    })
    .asPartial()
);

type Invoice = rt.Static<typeof Invoice>;

const Point = rt.Intersect(
  rt.Record({ x: rt.Number, y: rt.Number }),
  rt.Record({ label: rt.String }).asPartial()
);

type Point = rt.Static<typeof Point>;

const Line = rt.Intersect(
  rt.Record({ start: Point, end: Point }),
  rt.Record({ label: rt.String }).asPartial()
);

type Line = rt.Static<typeof Line>;
"
`;
